node/src/components/block_accumulator.rs

```diff
@@ -98,6 +108,11 @@ pub(crate) struct BlockAccumulator {
     recent_era_interval: u64,
     /// Tracks activity and assists with perceived tip determination.
     last_progress: Timestamp,
+    /// For each peer, a list of block hashes we first heard from them, and the timestamp when we
```

Comment:
This comment should state the invariant that they need to be in-order, I think.

Frankly, I think the logic is sufficiently complicated that this could have been spun-off as its own testable struct (newtype wrapping `VecDeque<(BlockHash, Timestamp)`).

Reply:
The comment says so, in the next line. ("from oldest to newest").

Didn't get a chance to try extracting the `VecDeque`-with-timestamps that enforces chronological order today. Since I have three approvals I'll merge this now, instead of asking everyone to re-review. If I have time, I can try it tomorrow in a separate PR.